### DISCUSSION TITLE
Unattended `TSInstall` possible for automated installations?

### DISCUSSION THREAD

**Discussion by miguno** [nvim-treesitter/nvim-treesitter#2725] [Q&A] [answered]
I'd like to install supported languages via a script, think:

```sh
nvim --headless "+TSInstall rust"
```

This doesn't seem to work because:

1. The command above does not terminate when installing a language that was not yet installed:

```sh
# First time the language 'rust' is installed
$ nvim --headless "+TSInstall rust"
[nvim-treesitter] [0/1] Downloading...[nvim-treesitter] [0/1] Creating temporary directory[nvim-treesitter] [0/1] Extracting...[nvim-treesitter] [0/1] Compiling...[nvim-treesitter] [1/1] Treesitt...arser for rust has been installed
# ^ does not terminate
```

2. There does not seem to be a way to auto-answer the confirmation question when attempting to install a language that was already installed:

```sh
# 'rust' is already installed
$ nvim --headless "+TSInstall rust"
rust parser already available: would you like to reinstall ? y/n: rust parser already available: would you like to reinstall ? y/n:
```

A simple attempt with `yes` didn't work, it resulted in the same behavior as the previous output.

```sh
$ yes | nvim --headless "+TSInstall rust"
rust parser already available: would you like to reinstall ? y/n: rust parser already available: would you like to reinstall ? y/n:
```

What's the recommended way to do this? The goal is to integrate the setup of treesitter into my fully automated dotfiles setup.

**Comment by kyazdani42** [comment]
try `TSInstallSync`

**Comment by miguno** [comment, marked as answer]
Thanks for replying!

`TSInstallSync` doesn't work either:

```sh
# Same with `yes | nvim --headless "+TSInstallSync! rust"`.
$ nvim --headless "+TSInstallSync! rust"
Downloading...
Creating temporary directory
Extracting...
Compiling...
Treesitter parser for rust has been installed
Press ENTER or type command to continue
```
^ hangs at this point.  FWIW, even when I manually press ENTER here, the prompt doesn't return.

I also tried with `TSUpdateSync`, but no luck:

```
# Same with `yes | nvim --headless "+TSUpdateSync rust"`
$ nvim --headless "+TSUpdateSync rust"
Parsers are up-to-date!
```
^ hangs at this point

**Reply by theHamsta** [reply]
You need to quit nvim after the installation has finished `nvim --headless "+TSInstallSync! rust" +qa`

**Reply by miguno** [reply]
Doh, you are totally right! Thank you.

One final question: For my use case, which is "Install latest parser XYZ if missing, otherwise update parser XYZ", I suppose `TSUpdateSync` is sufficient, rather than forcing re-installation with `TSInstallSync!`? The docs suggest it will also automatically install a parser if it is missing.

**Reply by theHamsta** [reply]
yes, TSUpdateSync installs or updates whatever is necessary.

**Reply by miguno** [reply]
Great, thanks again. Appreciate the quick responses! 👍
